src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1079 @@
+                    diff,
+                    ..
+                } => {
+                    soft_assert_eq!(diff, 1);
+                    epoch
+                }
+                _ => unreachable!("checked above"),
+            }
+        } else {
+            Epoch::new(1).expect("known to be non-zero")
+        };
+        // Note only writable catalogs attempt to increment the epoch.
+        if matches!(mode, Mode::Writable) {
+            epoch = Epoch::new(epoch.get() + 1).expect("known to be non-zero");
+        }
+
+        let mut catalog = PersistCatalogState {
+            mode,
+            persist_handle: self,
+            upper,
+            epoch,
+            snapshot: Snapshot::empty(),
+            fenced: false,
+        };
+
+        let mut txn = if is_initialized {
+            if !read_only {
+                let version =
+                    version.ok_or(CatalogError::Durable(DurableCatalogError::Uninitialized))?;
+                if version != STASH_VERSION {
+                    // TODO(jkosh44) Implement migrations.
+                    panic!("the persist catalog does not know how to perform migrations yet");
+                }
+            }
+
+            catalog.apply_updates(initial_snapshot.into_iter())?;
+            let mut txn = catalog.transaction().await?;
+            if let Some(deploy_generation) = deploy_generation {
+                txn.set_config(DEPLOY_GENERATION.into(), deploy_generation)?;
+            }
+            txn
+        } else {
+            // Get the current timestamp so we can record when we booted. We don't have to worry
+            // about `boot_ts` being less than a previously used timestamp because the catalog is
+            // uninitialized and there are no previous timestamps.
+            let boot_ts = now();
+
+            let mut txn = catalog.transaction().await?;
+            initialize::initialize(&mut txn, bootstrap_args, boot_ts, deploy_generation).await?;
+            txn
+        };
+
+        if !read_only {
+            add_new_builtin_clusters_migration(&mut txn)?;
+            add_new_builtin_cluster_replicas_migration(&mut txn, bootstrap_args)?;
+        }
+
+        if read_only {
+            let (txn_batch, _) = txn.into_parts();
+            // The upper here doesn't matter because we are only apply the updates in memory.
+            let updates = StateUpdate::from_txn_batch(txn_batch, catalog.upper);
+            catalog.apply_updates(updates.into_iter())?;
+        } else {
+            txn.commit().await?;
+        }
+
+        Ok(catalog)
+    }
+
+    /// Fetch the current upper of the catalog state.
+    // TODO(jkosh44) This isn't actually guaranteed to be linearizable. Before enabling this in
+    //  production we need a new linearizable solution.
+    async fn current_upper(&mut self) -> Timestamp {
+        self.write_handle
+            .fetch_recent_upper()
+            .await
+            .as_option()
```

Comment:
nit: `.as_option.cloned()` -> `.into_option()`

Reply:
I get the following error when I try that:

```
error[E0507]: cannot move out of a shared reference
   --> src/catalog/src/persist.rs:392:9
    |
392 | /         self.write_handle
393 | |             .fetch_recent_upper()
394 | |             .await
395 | |             .into_option()
    | |              ------------^
    | |______________|___________|
    |                |           move occurs because value has type `Antichain<mz_repr::Timestamp>`, which does not implement the `Copy` trait
    |                value moved due to this method call
    |
note: `Antichain::<T>::into_option` takes ownership of the receiver `self`, which moves value
   --> /home/joe/.cargo/git/checkouts/timely-dataflow-70b80d81d6cabd62/53adc68/timely/src/progress/frontier.rs:245:28
    |
245 |     pub fn into_option(mut self) -> Option<T> {
    |                            ^^^^
help: you can `clone` the value and consume it, but this might not be your desired behavior
    |
395 |             .clone().into_option()
    |              ++++++++
```

I think because `into_option()` consumes `self`, but `fetch_recent_upper()` returns a reference.

Reply:
Ah sorry! I misremembered it as returning an owned value